File: bit_login/sso/exceptions.py
from typing import Optional


class BitSsoError(Exception):
    """Base error raised by this package."""


class ConfigurationError(BitSsoError):
    """The SSO page did not contain the required login configuration."""


class CaptchaError(BitSsoError):
    """A captcha could not be obtained or recognized."""


class SmsVerificationError(BitSsoError):
    """The SMS code was rejected before login submission."""


class LoginError(BitSsoError):
    """CAS returned the login page with an error."""

    def __init__(
        self,
        message: str,
        *,
        code: str = "",
        final_url: str = "",
        status_code: Optional[int] = None,
        redirect_count: int = 0,
        ticket_issued: bool = False,
        risk_mode: str = "",
        flow_replaced: Optional[bool] = None,
        captcha_required: bool = False,
    ) -> None:
        display_message = f"{message} (code: {code})" if code else message
        diagnostics = []
        if status_code is not None:
            diagnostics.append(f"status={status_code}")
        diagnostics.append(f"redirects={redirect_count}")
        diagnostics.append(f"risk={risk_mode or 'unknown'}")
        diagnostics.append(f"ticket={'yes' if ticket_issued else 'no'}")
        flow_state = "unknown" if flow_replaced is None else (
            "replaced" if flow_replaced else "same"
        )
        diagnostics.append(f"flow={flow_state}")
        diagnostics.append(f"captcha={'yes' if captcha_required else 'no'}")
        display_message += " [" + ", ".join(diagnostics) + "]"
        super().__init__(display_message)
        self.code = code
        self.final_url = final_url
        self.status_code = status_code
        self.redirect_count = redirect_count
        self.ticket_issued = ticket_issued
        self.risk_mode = risk_mode
        self.flow_replaced = flow_replaced
        self.captcha_required = captcha_required

File: bit-login/src/commonMain/kotlin/cn/bit101/bitlogin/sso/Exceptions.kt
package cn.bit101.bitlogin.sso

open class BitSsoError(message: String, cause: Throwable? = null) : Exception(message, cause)

class ConfigurationError(message: String, cause: Throwable? = null) : BitSsoError(message, cause)

class CaptchaError(message: String, cause: Throwable? = null) : BitSsoError(message, cause)

class SmsVerificationError(message: String, cause: Throwable? = null) : BitSsoError(message, cause)

/**
 * Mirrors Python `requests.HTTPError` raised by `response.raise_for_status()`.
 * Carries the HTTP status so SsoLogin can translate it to the same Chinese
 * messages the Python wrapper produces (429 / 5xx / other 4xx).
 */
class SsoHttpException(
    val statusCode: Int,
    val url: String,
    val bodyText: String,
) : BitSsoError("HTTP $statusCode from $url")

class LoginError(
    message: String,
    val code: String = "",
    val finalUrl: String = "",
    val statusCode: Int? = null,
    val redirectCount: Int = 0,
    val ticketIssued: Boolean = false,
    val riskMode: String = "",
    val flowReplaced: Boolean? = null,
    val captchaRequired: Boolean = false,
) : BitSsoError(
    buildString {
        append(message)
        if (code.isNotEmpty()) append(" (code: ").append(code).append(')')
        append(" [")
        statusCode?.let { append("status=").append(it).append(", ") }
        append("redirects=").append(redirectCount)
        append(", risk=").append(riskMode.ifEmpty { "unknown" })
        append(", ticket=").append(if (ticketIssued) "yes" else "no")
        append(", flow=").append(
            when (flowReplaced) {
                true -> "replaced"
                false -> "same"
                null -> "unknown"
            },
        )
        append(", captcha=").append(if (captchaRequired) "yes" else "no")
        append(']')
    },
)

/**
 * Mirrors Python `_LOGIN_ERROR_MESSAGES` in `client.py`. Maps CAS error codes
 * to human-readable Chinese messages.
 */
val LOGIN_ERROR_MESSAGES: Map<String, String> = mapOf(
    "1030027" to "用户名或密码错误",
    "1030028" to "账号已被锁定",
    "1030031" to "用户名或密码错误",
    "1320007" to "验证码错误或已失效",
    "1320010" to "图形验证码错误",
    "1330001" to "登录被账号风控拒绝",
    "1410040" to "账号状态无效",
    "1410041" to "账号状态无效",
    "3910001" to "账号已休眠，请先完成账号激活",
)

/**
 * Mirrors Python `_format_login_error`. Returns (message, code) tuple.
 *
 * - If `message` is empty but looks numeric, swap it into `code`.
 * - If `message` is non-empty, use it.
 * - If `code` is non-empty, look up [LOGIN_ERROR_MESSAGES] (default "CAS login failed").
 * - If `ticketIssued`, return the ticket-bounce message.
 * - If `riskMode == "error-fallback"`, return the USTC risk failure message.
 * - Otherwise return the generic execution-expired message.
 */
fun formatLoginError(
    rawMessage: String,
    rawCode: String,
    ticketIssued: Boolean,
    riskMode: String,
): Pair<String, String> {
    var message = rawMessage.trim()
    var code = rawCode.trim()
    if (code.isEmpty() && message.isNotEmpty() && message.all { it.isDigit() }) {
        code = message
        message = ""
    }
    if (message.isNotEmpty()) return message to code
    if (code.isNotEmpty()) return (LOGIN_ERROR_MESSAGES[code] ?: "CAS login failed") to code
    if (ticketIssued) return "CAS issued a service ticket, but the service redirected back to the login page" to ""
    if (riskMode == "error-fallback") return "CAS returned the login page without an error code after USTC risk-token acquisition failed" to ""
    return "CAS returned the login page without an error code; the execution may have expired or risk control may have rejected the request" to ""
}
